### PR TITLE
v2.final

### PR DESCRIPTION
**Pull request by nikfilippas**
WIP for CCLv3. CCLv3 branches to be based on here. This was originally a clone of `master` with an internal restructure in `halos/` and `base/` as specified below:

- [x] Declutter `halos/profiles`: each profile in its own module.
- [x] Declutter `halos/concentration`: each concentration in its own module.
- [x] Declutter `halos/hmfunc`: each mass function in its own module.
- [x] Declutter `halos/hbias`: each halo bias in its own module.
- [x] Declutter `halos/halo_model`: separate modules for `pk_1pt`, `pk_2pt`, `pk_4pt`.
- [x] Declutter `base.py`: split into `caching.py`, `deprecations.py` `ccl_object.py` and moved these under `pyccl/base/`.

List of things to change:
- [x] baryons (#1039 )
- [x] halos (#1043 , #1068 , #1069 , #1072  )
- [x] nl_pt (#1056 , #1073  )
- [x] background (#1041 )
- [x] boltzmann (#1041 )
- [x] cells (#1041 )
- [x] cosmology (#1071 )
- [x] correlations (#1042 )
- [x] covariances (#1046 )
- [x] neutrinos (#1047 , #1049 , #1058 )
- [x] pk2d (#1048 )
- [x] power (#1047 )
- [x] tk3d (#1048 )
- [x] tracers (#1040 )
- [ ] Fix docstrings (bring them in from [myccl](https://github.com/nikfilippas/myccl/)). (#1053 )
- [x] tests (#1059  )

To remember:
- [x] reinstate `mass_def` in `halos` (#1077 )

Nick's to-do list, once everything is merged in here: (#1075 )

- [ ] Go through [`Issues`](https://github.com/LSSTDESC/CCL/issues) and [`Milestones`](https://github.com/LSSTDESC/CCL/milestone/5) and see what can be linked to here. (remaining ones in #1053 )
- [x] Assert statements in tests only with `pytest`. (#1059   )
- [x] Restructure `star` to new dir-tree, and temporarily rebase `v3_wip --> star` to check for any leftover things to change.
- [x] Bring back old tests to see that everything works and API is preserved. (#1077 )
- [x] Do a final pass and examine code quality. (#1075 )
    - [x] simplify internal imports with `__all__` and `*`
    - [x] if-else return statements
    - [x] `return` without defining extra variable
    - [x] `self._one_over_4_pi` replace the like with numbers
    - [x] remove all useless type-checking (type hints where needed to enable use with `mypy`)
    - [x] methods of cosmology - no import needed

### CODE REVIEW

File: pyccl/halos/concentration/__init__.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+from .ishiyama21 import ConcentrationIshiyama21
+from .klypin11 import ConcentrationKlypin11
+from .prada12 import ConcentrationPrada12
+
+
+__all__ = (
+    "Concentration", "concentration_from_name",
+    "ConcentrationBhattacharya13",
+    "ConcentrationConstant",
+    "ConcentrationConstant",
```
Comment (damonge) [inline pyccl/halos/concentration/__init__.py:15]: repeated

**Review by damonge — COMMENTED** [review]
Just one typo

**Comment by nikfilippas** [comment]
Coverage decreased slightly but not to worry, this is largely because we aren't testing `warn_api`. We can add tests for this at the end.

**Review by damonge — APPROVED** [review]
Merging to master after last pure v2 release

**Pull request merged**